### PR TITLE
Split web serving, move mysql to commons

### PR DESCRIPTION
**Pull request by alok123t**

### CODE REVIEW

**Comment by VamshiMuthineni** [comment]
mysql and hhvm are not available for this workload. that's why they're not present.

**Comment by neo-apz** [comment]
At least MySQL seems to be available 🤔 
https://packages.debian.org/search?arch=arm64&keywords=mysql-server

and there is this blog post from three years ago that claims HHVM is available on ARM:
https://hhvm.com/blog/2017/03/09/how-the-cyber-elephant-got-his-arm.html

**Comment by neo-apz** [comment]
This is a 4-tier benchmark; if any of the tiers can be available on other platforms, then they should available.

**Comment by VamshiMuthineni** [comment]
mysql is only available on sid (we are using buster). maybe we can try adding to sources.list or add repository with tag sid, then it will try to fetch from sid and probably do a force install. (like the one we tried for java initially, before doing with adoptopenjdk). Can I proceed this way?

For memcached, we can add arm64 in .travis.yml as the base image is already available to arm.

and For hhvm it's quite complicated. We've to build from source and I faced a lot of issues while doing so. When raised a ticket for a particular issue in hhvm git, I got the response that they no longer support arm. [https://github.com/facebook/hhvm/issues/8703](url)

**Comment by neo-apz** [comment]
I see.

OK please changed the files for Memcached and MySQL as you suggested, and we can have the web-server without HHVM on ARM for now.

**Comment by VamshiMuthineni** [comment]
support for faban and memcached are added. But for mysql if I install via above suggested method facing some connection issues. Maybe, we can update mysql later.

**Comment by pradeepk5** [comment]
Thanks for the updates.
When testing the benchmarks I get the following errors in the client when the client tries to create users.

```
Starting Faban Server                                                                                                                                                           
Please point your browser to http://iccluster101:9980/                                                      
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -5         
        at java.lang.AbstractStringBuilder.substring(AbstractStringBuilder.java:935)                        
        at java.lang.StringBuilder.substring(StringBuilder.java:76)                                         
        at setup.UserGenerator.createUsers(UserGenerator.java:137)                                                                
        at setup.UserGenerator.main(UserGenerator.java:147)                                   
Buildfile: /web20_benchmark/build.xml   

```

**Comment by pradeepk5** [comment]
> Thanks for the updates.
> When testing the benchmarks I get the following errors in the client when the client tries to create users.
> 
> ```
> Starting Faban Server                                                                                                                                                           
> Please point your browser to http://iccluster101:9980/                                                      
> Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -5         
>         at java.lang.AbstractStringBuilder.substring(AbstractStringBuilder.java:935)                        
>         at java.lang.StringBuilder.substring(StringBuilder.java:76)                                         
>         at setup.UserGenerator.createUsers(UserGenerator.java:137)                                                                
>         at setup.UserGenerator.main(UserGenerator.java:147)                                   
> Buildfile: /web20_benchmark/build.xml   
> ```

Here are some debug infromation which might help.

Error is happening at the last line of this block where it tries to substring the response.
```
sb = http.fetchURL(hostURL+"/action/useradd", postRequest, headers);
			int startIndex = sb.indexOf("GUID#")+"GUID#".length();
			int endIndex = sb.indexOf("#", startIndex);
			String guid = sb.substring(startIndex, endIndex);
```
Stack trace
```
main[1] where                                                                    
  [1] java.lang.AbstractStringBuilder.substring (AbstractStringBuilder.java:935)   
  [2] java.lang.StringBuilder.substring (StringBuilder.java:76)                                                                                            
  [3] setup.UserGenerator.createUsers (UserGenerator.java:137)                           
  [4] setup.UserGenerator.main (UserGenerator.java:147)   
```

Here are the local variables at time of failure
```
Local variables:
i = 27
http = instance of com.sun.faban.driver.transport.sunhttp.SunHttpTransport(id=900)
sb = instance of java.lang.StringBuilder(id=901)
loginPostRequest = "__elgg_token=
<html>
        <head>
                <title>Fatal Error.</title>
                <meta http-equiv=&__elgg_ts=tml>
<html>
        <head>
                <title>Fatal Error.</title>
                <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />

                <style type="text/css">

                body {
                        text-align:left;
                        margin:0;
                        padding:0;
                        background: #4690d6;
                        font: 80%/1.5  "Lucida Grande"&username=admin&password=admin1234"
headers = instance of java.util.HashMap(id=903)
user = instance of setup.UserEntity(id=904)
postRequest = "__elgg_token=0
                             ='??s?[?k???`h?]l?n?????BeI?????>Iv?~L?l?&__elgg_ts=SMo?0
                                                                                      ='??s?[?k???`h?]l?n?????BeI?????>Iv?~L?l?"??????????n??^?????U2???f       ???3bn??vYe?/??H??I?? ??V????????&????+h:f?v?C?I ?????D ?57?%????}?&name=WnZMKowUea&username=WnZMKowUea&email=WnZMKowUea@gmail.com&password=Dp4qAd3coA&password2=Dp4qAd3coA&admin=0"
startIndex = 4
endIndex = -1

```
startIndex looking for `"GUID#` in the response, and it is not found in the `sb`. 
This happens randomly, which means for some user profile generation the didn't occur.

**Comment by alok123t** [comment]
@pradeepk5 I reverted the MySQL version to the previous working version of 5.5, the web_server also uses this version. Tested by launching each tier on different VMs and the above error is no longer present.

**Comment by pradeepk5** [comment]
> @pradeepk5 I reverted the MySQL version to the previous working version of 5.5, the web_server also uses this version. Tested by launching each tier on different VMs and the above error is no longer present.

Thanks for the update.

File: benchmarks/web-serving/web_server/Dockerfile [hunk 1]
```diff
@@ -5,8 +5,8 @@ USER root
```
Comment (pradeepk5) [inline benchmarks/web-serving/web_server/Dockerfile:5]: This should use the common debian base image. Could you please change this to the debian os?

File: commons/mysql/Dockerfile [hunk 2]
```diff
@@ -2,29 +2,25 @@ FROM ubuntu:14.04
 LABEL maintainer="Mark Sutherland <mark.sutherland@epfl.ch>"
```
Comment (pradeepk5) [inline commons/mysql/Dockerfile:2]: This should be changed to the debian base os.

File: benchmarks/web-serving/web_server/Dockerfile (hunk 1 above)
Comment (alok123t) [reply benchmarks/web-serving/web_server/Dockerfile:5]: php5-* are only available on older Debian versions, check [link](https://packages.debian.org/search?suite=all&arch=any&mode=path&searchon=names&keywords=php5)

File: commons/mysql/Dockerfile (hunk 2 above)
Comment (alok123t) [reply commons/mysql/Dockerfile:2]: MySQL v5.5 doesn't seem to be supported on Debian buster, check [link](http://repo.mysql.com/apt/debian/dists/buster/) 
Also, this [mysql-dump](https://github.com/parsa-epfl/cloudsuite/blob/dev/benchmarks/web-serving/db_server/files/elgg_db.dump) is for v5.5. We need to test if it gets imported successfully on later versions.

File: benchmarks/web-serving/web_server/Dockerfile (hunk 1 above)
Comment (neo-apz) [reply benchmarks/web-serving/web_server/Dockerfile:5]: @alok123t thanks for the update. Is there a specific reason we should still stick to php5? It's quite old and the support for that version has ended. Why not upgrade it to a more recent version of php?
The same goes for MySQL; v5.5 is quite old and it's not supported anymore. Can we have both of these upgraded?

File: commons/mysql/Dockerfile (hunk 2 above)
Comment (neo-apz) [reply commons/mysql/Dockerfile:2]: @alok123t this PR looks fine and can be merged.
However, I think it's better if we can bump up the PHP and MySQL versions, as both are currently very old. Now that we are upgrading these benchmarks and we plan to have a new release, it makes sense to also upgrade these. This also resolves the issue with packages not being available for the base OS and prevents other issues arising in the near future. Can you do this?

**Comment by neo-apz** [comment]
@VamshiMuthineni @ajaygk95 @komal1204
This PR looks fine and can be merged.
However, I think it's better if we can bump up the PHP and MySQL versions, as both are currently very old (not maintained and obsolete). Now that we are upgrading these benchmarks and we plan to have a new release, it makes sense to also upgrade these. This also resolves the issue with packages not being available for the base OS and prevents other issues arising in the near future.
Can you do this?

**Comment by komal1204** [comment]
> @VamshiMuthineni @ajaygk95 @komal1204
> This PR looks fine and can be merged.
> However, I think it's better if we can bump up the PHP and MySQL versions, as both are currently very old (not maintained and obsolete). Now that we are upgrading these benchmarks and we plan to have a new release, it makes sense to also upgrade these. This also resolves the issue with packages not being available for the base OS and prevents other issues arising in the near future.
> Can you do this?

Hello,
yes, I am working on upgrading both MySQL and PHP; testing with PHP7.3 is almost done, however with updated MySQL, I am getting some errors, so once those are fixed, will update this PR.

**Comment by komal1204** [comment]
Hello @yunho-oh ,

This PR can be closed as the changes needed were done in pr #267 , which is already merged into dev.

Thanks,
Komal

**Comment by yunho-oh** [comment]
Done

**Comment by komal1204** [comment]
Hello ,

pr #267  just addresses moving and updating MySQL( now using MariaDb) to commons

PR #267 does not include changes for upgrading PHP.

I tried updating php to php 7.3 by making changes in the webserving webserver's Dockerfile, bootstrap.sh and nginx_sites_avail.append , and the server builds  correctly.

However when I try to run the faban client, I get the following error - 

Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -16
	at java.lang.AbstractStringBuilder.substring(AbstractStringBuilder.java:935)
	at java.lang.StringBuilder.substring(StringBuilder.java:76)
	at setup.UserGenerator.updateElggTokenAndTs(UserGenerator.java:39)
	at setup.UserGenerator.createUsers(UserGenerator.java:103)
	at setup.UserGenerator.main(UserGenerator.java:147)
Buildfile: /web20_benchmark/build.xml

**Pull request closed without merging**